jerry-ext/include/jerryscript-ext/context.h

```diff
@@ -0,0 +1,58 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_CONTEXT_H
+#define JERRYX_CONTEXT_H
+
+#include "jerryscript.h"
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its second parameter
+ */
+void *jerryx_context_init (void);
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its third parameter
+ */
+void jerryx_context_deinit (void *user_context_p);
+
+typedef struct
+{
+  /** the initialization function */
+  jerry_user_context_init_cb init_cb;
+  /** the deinitialization function */
+  jerry_user_context_deinit_cb deinit_cb;
+} jerryx_context_slot_t;
+
+#ifdef __MACH__
+#define JERRYX_CONTEXT_SECTION "__DATA,jerryx_ctx_slots"
+#else /* !__MACH__ */
+#define JERRYX_CONTEXT_SECTION "jerryx_ctx_slots"
+#endif /* __MACH__ */
+
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)         \
+static jerryx_context_slot_t __jerryx_context_slot_ ## slot_name                  \
+__attribute__((used, section(JERRYX_CONTEXT_SECTION), aligned(sizeof(void *)))) = \
+{                                                                                 \
+  .init_cb = (init_cb_ptr),                                                       \
+  .deinit_cb = (deinit_cb_ptr)                                                    \
+};
```

Comment:
I know it seems unnatural for there to be no semicolon after what looks like a statement, but again, if we need to implement this using a constructor, then that's a function, and no semicolon may follow the closing brace of a function.

Look at https://github.com/nodejs/node/blob/master/src/node.h#L456-L473. The macro there must not have a semicolon after it when used because it's implemented as a constructor. The reason it's implemented as a constructor is because Node modules are dlopen()ed so they're not in any section at boot time.

If we (sooner or) later wish to support dynamic loading, we should make it possible with a simple recompile, rather than code change + recompile.